### PR TITLE
[llvm-exegesis] Add tablegen support for validation counters

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch adds support in the llvm-exegesis tablegen emitter for validation counters. Full support for validation counters in llvm-exegesis is added in a future patch.

### CODE REVIEW

File: llvm/include/llvm/Target/TargetPfmCounters.td [hunk 1]
```diff
@@ -28,6 +28,22 @@ class PfmIssueCounter<string resource_name, string counter>
   string ResourceName = resource_name;
 }
 
+class ValidationEvent <int event_number> {
```
Comment (legrosbuffle) [inline llvm/include/llvm/Target/TargetPfmCounters.td:31]: This needs doc to explain what a `ValidationEvent` and `PfmValidationCounter` are.

File: llvm/tools/llvm-exegesis/lib/Target.h [hunk 2]
```diff
@@ -59,6 +68,8 @@ struct PfmCountersInfo {
   const IssueCounter *IssueCounters;
   unsigned NumIssueCounters;
 
+  std::unordered_map<ValidationEvent, const char *> ValidationCounters;
```
[the remark this answers is not in the document]
Comment (boomanaiden154) [reply llvm/tools/llvm-exegesis/lib/Target.h:71]: Good point. I've refactored the emitter to emit a sorted array of pairs, like you suggested.

File: llvm/include/llvm/Target/TargetPfmCounters.td (hunk 1 above)
Comment (boomanaiden154) [reply llvm/include/llvm/Target/TargetPfmCounters.td:31]: Added additional comments on what these are for. Please let me know if they don't clarify things enough.

**Comment by boomanaiden154** [comment]
@legrosbuffle When you get a chance, are you able to take another look at this along with the other patches in the stack?
1. https://github.com/llvm/llvm-project/pull/76653
2. https://github.com/llvm/llvm-project/pull/76788

I also have a couple other patches that would be nice to land soon. It would be great if you could take a look at them also:
1. https://github.com/llvm/llvm-project/pull/77226
2. https://github.com/llvm/llvm-project/pull/77224
3. https://github.com/llvm/llvm-project/pull/77020
4. https://github.com/llvm/llvm-project/pull/76895

File: llvm/include/llvm/Target/TargetPfmCounters.td [hunk 3]
```diff
@@ -28,6 +28,27 @@ class PfmIssueCounter<string resource_name, string counter>
   string ResourceName = resource_name;
 }
 
+// Definition of a validation event. A validation event represents a specific
+// event that can be meaasured using performance counters that is interesting
```
Comment (legrosbuffle) [inline llvm/include/llvm/Target/TargetPfmCounters.td:32]: typo: meaasured

File: llvm/utils/TableGen/ExegesisEmitter.cpp [hunk 4]
```diff
@@ -100,6 +105,17 @@ ExegesisEmitter::ExegesisEmitter(RecordKeeper &RK)
   Target = std::string(Targets[0]->getName());
 }
 
+struct ValidationCounterInfo {
+  int64_t EventNumber;
+  StringRef EventName;
+  unsigned PfmCounterID;
+};
+
+bool CompareValidationCounterInfo(ValidationCounterInfo &LHS,
```
Comment (legrosbuffle) [inline llvm/utils/TableGen/ExegesisEmitter.cpp:114]: [nit] `EventNumberLess` ?

Also, `const ValidationCounterInfo&`.

**Review by legrosbuffle — APPROVED** [review]

**Pull request merged**